src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp

```diff
@@ -0,0 +1,712 @@
+  z = 0.5 *
+      ((radius_one_ * cos(theta_max_one_ * z) + center_one_[2]) * (1.0 - zbar) +
+       center_two_[2] * (1.0 + zbar));
+  return target_coords;
+}
+
+std::optional<std::array<double, 3>> UniformCylindricalFlatEndcap::inverse(
+    const std::array<double, 3>& target_coords) const {
+  // First do some easy checks that target_coords is in the range
+  // of the map.  Need to accept points that are out of range by roundoff.
+
+  // check z >= z_plane_one_
+  if (target_coords[2] < z_plane_one_ and
+      not equal_within_roundoff(target_coords[2], z_plane_one_)) {
+    return std::nullopt;
+  }
+
+  // check that target point is at smaller z than circle_two
+  if (target_coords[2] > center_two_[2] and
+      not equal_within_roundoff(target_coords[2], center_two_[2])) {
+    return std::nullopt;
+  }
+
+  // check that point is outside or on sphere_one_
+  const double r_minus_center_one_squared =
+      square(target_coords[0] - center_one_[0]) +
+      square(target_coords[1] - center_one_[1]) +
+      square(target_coords[2] - center_one_[2]);
+  if (r_minus_center_one_squared < square(radius_one_) and
+      not equal_within_roundoff(sqrt(r_minus_center_one_squared),
+                                radius_one_)) {
+    // sqrt above because we don't want to change the scale of
+    // equal_within_roundoff too much.
+    return {};
+  }
+
+  // Check if the point is inside the cone
+  const double lambda_tilde =
+      (target_coords[2] - z_plane_one_) / (center_two_[2] - z_plane_one_);
+  const double rho_tilde =
+      sqrt(square(target_coords[0] - center_one_[0] -
+                  lambda_tilde * (center_two_[0] - center_one_[0])) +
+           square(target_coords[1] - center_one_[1] -
+                  lambda_tilde * (center_two_[1] - center_one_[1])));
+  const double circle_radius =
+      radius_one_ * sin(theta_max_one_) * (1.0 - lambda_tilde) +
+      radius_two_ * lambda_tilde;
+  if (rho_tilde > circle_radius and
+      not equal_within_roundoff(rho_tilde, circle_radius,
+                                std::numeric_limits<double>::epsilon() * 100.0,
+                                radius_two_)) {
+    return {};
+  }
+
+  // To find rhobar we will do numerical root finding.
+  // function_to_zero is the function that is zero when rhobar has the
+  // correct value.
+  const auto function_to_zero = [this, &target_coords](const double rhobar) {
+    return this_function_is_zero_for_correct_rhobar(
+        rhobar, center_one_, center_two_, radius_one_, radius_two_,
+        theta_max_one_, target_coords);
+  };
+
+  // To find rhobar we will do numerical root finding.
+  // function_and_deriv_to_zero returns a std::pair consisting of the
+  // function that is zero when rhobar has the correct value, and the
+  // derivative of that function with respect to rhobar.
+  const auto function_and_deriv_to_zero = [this, &target_coords](
+                                              const double rhobar) {
```

Comment:
Why doesn't this function just call `this_function_is_zero_for_correct_rhobar`?

Reply:
I just removed this function because it is no longer needed as-is.  Only a function that computes the deriv is needed, which I added instead of this function.